drivers/adc/adc_ad4114.c

```diff
@@ -0,0 +1,457 @@
+
+	uint8_t buffer_tx[6] = {0};
+	uint8_t buffer_rx[ARRAY_SIZE(buffer_tx)] = {0xFF};
+	const struct spi_buf tx_buf[] = {{
+		.buf = buffer_tx,
+		.len = ARRAY_SIZE(buffer_tx),
+	}};
+	const struct spi_buf rx_buf[] = {{
+		.buf = buffer_rx,
+		.len = ARRAY_SIZE(buffer_rx),
+	}};
+	const struct spi_buf_set tx = {
+		.buffers = tx_buf,
+		.count = ARRAY_SIZE(tx_buf),
+	};
+	const struct spi_buf_set rx = {
+		.buffers = rx_buf,
+		.count = ARRAY_SIZE(rx_buf),
+	};
+	buffer_tx[0] = AD4114_CMD_READ | reg_addr;
+
+	ret = spi_transceive_dt(&config->spi, &tx, &rx);
+	if (ret != 0) {
+		LOG_ERR("Read reg: error writing register 0x%X (%d)", reg_addr, ret);
+		return ret;
+	}
+
+	// Copy received data in output buffer
+	for (uint8_t index = 0; index < reg_size; index++) {
+		buffer[index] = buffer_rx[index + 1];
+	}
+
+	return ret;
+}
+
+static void adc_context_start_sampling(struct adc_context *ctx)
+{
+	struct adc_ad4114_data *data = CONTAINER_OF(ctx, struct adc_ad4114_data, ctx);
+
+	data->channels = ctx->sequence.channels;
+	data->repeat_buffer = data->buffer;
+
+	k_sem_give(&data->sem);
+}
+
+static void adc_context_update_buffer_pointer(struct adc_context *ctx, bool repeat_sampling)
+{
+	struct adc_ad4114_data *data = CONTAINER_OF(ctx, struct adc_ad4114_data, ctx);
+
+	if (repeat_sampling) {
+		data->buffer = data->repeat_buffer;
+	}
+}
+
+static int adc_ad4114x_validate_buffer_size(const struct device *dev,
+					    const struct adc_sequence *sequence)
+{
+	uint8_t channels;
+	size_t needed;
+
+	channels = POPCOUNT(sequence->channels);
+	needed = channels * sizeof(uint32_t);
+
+	if (sequence->buffer_size < needed) {
+		LOG_ERR("insufficient buffer size");
+		return -ENOMEM;
+	}
+
+	return 0;
+}
+
+static int adc_ad4114_start_read(const struct device *dev, const struct adc_sequence *sequence)
+{
+	struct adc_ad4114_data *data = dev->data;
+	const struct adc_ad4114_config *config = dev->config;
+	int ret;
+	uint8_t write_reg[2];
+
+	ret = adc_ad4114x_validate_buffer_size(dev, sequence);
+	if (ret < 0) {
+		LOG_ERR("insufficient buffer size");
+		return ret;
+	}
+
+	data->channels_cfg = sequence->channels;
+	for (uint32_t i = 0U; i < AD4114_CHAN_NUMBER; i++) {
+		if ((BIT(i) & sequence->channels) != 0) {
+			write_reg[0] = 0x80 | (uint8_t)((config->map_input[i] >> 8) & 0xFF);
+			write_reg[1] = (uint8_t)(config->map_input[i] & 0xFF);
+			LOG_DBG("Enable channel %d with mapping %X %X, raw %X", i, write_reg[0],
+				write_reg[1], config->map_input[i]);
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		} else {
+			LOG_DBG("Disable channel %d", i);
+			write_reg[0] = 0x0;
+			write_reg[1] = 0x0;
+			ad4114_write_reg(dev, AD4114_CHANNEL_0_REG + i, write_reg, 2);
+		}
+	}
+
+	// Configure the buffer
+	data->buffer = sequence->buffer;
+
+	// Wait to acquire first samples
+	// This is enough with default filter paramaters
+	// If settling is higer than 200us, first sample can be read to 0
+	k_usleep(200);
```

Comment:
Can't you use `IFMODE.DATA_STAT` and `STATUS.RDY` to check if data is actually valid?

Reply:
I have changed the logic to wait a not ready flag that indicates that the conversion has started with the new configuration. Now the setup is independent of the data rate.